### ISSUE TITLE
[Android 8.x] App sometimes force closes on opening Widgets tab

### ISSUE THREAD

**Issue by SanchithHegde** [jahirfiquitiva/Kuper#24]
- [x] I have verified there are no duplicate active or recent bugs, questions, or requests.
- [x] I have verified that I am using the latest version of Kuper.

### Device/dashboard info:
 - Kuper version: `1.3.1`
 - Android version: `8.1`
 - Device Manufacturer: `Motorola`
 - Device Name: `Moto G3 (2015)`

### Reproduction Steps
*This issue occurs only on devices with Andoid 8.x. I have tested this on devices with Android 7.x and 5.x also, and I did not experience any issues.*
1. Open app.
2. Open Widgets tab.

### Expected Result
App does not force close on opening Widgets tab.

### Actual Result
App ***sometimes*** force closes on opening Widgets tab.

### Log of App Force Closing
[[Android 8.x] Kuper Force Close Log](https://github.com/jahirfiquitiva/Kuper/files/1984030/Android.Oreo.Kuper.Force.Close.txt)
- *Log records the Kuper sample app force closing on opening the Widgets tab.*
- *Verbose level log, recorded using MatLog app.*

#### Notes
 - I didn't experience this issue on Kuper `v1.3.0`. I experienced this issue only after the `v1.3.1` update.
 - I didn't experience this issue every time. This happens sometimes only. This also happened after the first launch after installing the app.

**Comment by jahirfiquitiva** [comment]
Logs do not contain any useful info. Please make sure logs contain `jahirfiquitiva.libs.kuper` somewhere

**Comment by SanchithHegde** [comment]
Okay, will get back to you with proper logs.

**Comment by SanchithHegde** [comment]
I've got two more instances of the app force closing from a huge log dump. I've included only the relevant parts. Please find them linked below.
- https://del.dog/tirolebupu.sql
- https://del.dog/izodugefuk.sql

**Comment by jahirfiquitiva** [comment]
Ok, I'll fix this by putting the **actual** logs here. I will try to look into them this weekend, please be patient :sweat_smile: 

```groovy
05-07 23:08:05.440 W/System.err(31497): java.util.zip.ZipException: error reading zip file
05-07 23:08:05.441 W/System.err(31497): 	at java.util.zip.ZipFile.read(Native Method)
05-07 23:08:05.441 W/System.err(31497): 	at java.util.zip.ZipFile.-wrap0(Unknown Source:0)
05-07 23:08:05.441 W/System.err(31497): 	at java.util.zip.ZipFile$ZipFileInputStream.read(ZipFile.java:758)
05-07 23:08:05.441 W/System.err(31497): 	at java.util.zip.ZipFile$ZipFileInflaterInputStream.fill(ZipFile.java:447)
05-07 23:08:05.441 W/System.err(31497): 	at java.util.zip.InflaterInputStream.read(InflaterInputStream.java:163)
05-07 23:08:05.441 W/System.err(31497): 	at java.io.FilterInputStream.read(FilterInputStream.java:107)
05-07 23:08:05.441 W/System.err(31497): 	at kotlin.io.ByteStreamsKt.copyTo(Unknown Source:12)
05-07 23:08:05.441 W/System.err(31497): 	at jahirfiquitiva.libs.kuper.helpers.extensions.FileKt.copyFromTo(Unknown Source:32)
05-07 23:08:05.441 W/System.err(31497): 	at jahirfiquitiva.libs.kuper.providers.viewmodels.KuperViewModel.getWidgetPreviewsPathFromZip(Unknown Source:175)
05-07 23:08:05.441 W/System.err(31497): 	at jahirfiquitiva.libs.kuper.providers.viewmodels.KuperViewModel.internalLoad(Unknown Source:224)
05-07 23:08:05.441 W/System.err(31497): 	at jahirfiquitiva.libs.kuper.providers.viewmodels.KuperViewModel.internalLoad(Unknown Source:2)
05-07 23:08:05.441 W/System.err(31497): 	at jahirfiquitiva.libs.archhelpers.viewmodels.ItemViewModel.safeInternalLoad(Unknown Source:15)
05-07 23:08:05.441 W/System.err(31497): 	at jahirfiquitiva.libs.archhelpers.viewmodels.ItemViewModel.access$safeInternalLoad(Unknown Source:0)
05-07 23:08:05.441 W/System.err(31497): 	at jahirfiquitiva.libs.archhelpers.viewmodels.ItemViewModel$loadData$1.doLoad(Unknown Source:4)
05-07 23:08:05.441 W/System.err(31497): 	at jahirfiquitiva.libs.archhelpers.tasks.QAsync$execute$1.invoke(Unknown Source:23)
05-07 23:08:05.441 W/System.err(31497): 	at jahirfiquitiva.libs.archhelpers.tasks.QAsync$execute$1.invoke(Unknown Source:2)
05-07 23:08:05.441 W/System.err(31497): 	at org.jetbrains.anko.AsyncKt$doAsync$1.invoke(Unknown Source:4)
05-07 23:08:05.441 W/System.err(31497): 	at org.jetbrains.anko.AsyncKt$doAsync$1.invoke(Unknown Source:0)
05-07 23:08:05.441 W/System.err(31497): 	at org.jetbrains.anko.AsyncKt$sam$Callable$761a5578.call(Unknown Source:2)
05-07 23:08:05.441 W/System.err(31497): 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
05-07 23:08:05.441 W/System.err(31497): 	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:301)
05-07 23:08:05.441 W/System.err(31497): 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
05-07 23:08:05.441 W/System.err(31497): 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
05-07 23:08:05.441 W/System.err(31497): 	at java.lang.Thread.run(Thread.java:764)
05-07 23:08:05.442 D/skia    (31497): --- SkAndroidCodec::NewFromStream returned null
05-07 23:08:05.442 W/System.err(31497): java.lang.IllegalStateException: BitmapFactory.decodeFile(preview.absolutePath) must not be null
05-07 23:08:05.443 W/System.err(31497): 	at jahirfiquitiva.libs.kuper.providers.viewmodels.KuperViewModel.getWidgetPreviewsPathFromZip(Unknown Source:389)
05-07 23:08:05.443 W/System.err(31497): 	at jahirfiquitiva.libs.kuper.providers.viewmodels.KuperViewModel.internalLoad(Unknown Source:224)
05-07 23:08:05.443 W/System.err(31497): 	at jahirfiquitiva.libs.kuper.providers.viewmodels.KuperViewModel.internalLoad(Unknown Source:2)
05-07 23:08:05.443 W/System.err(31497): 	at jahirfiquitiva.libs.archhelpers.viewmodels.ItemViewModel.safeInternalLoad(Unknown Source:15)
05-07 23:08:05.443 W/System.err(31497): 	at jahirfiquitiva.libs.archhelpers.viewmodels.ItemViewModel.access$safeInternalLoad(Unknown Source:0)
05-07 23:08:05.443 W/System.err(31497): 	at jahirfiquitiva.libs.archhelpers.viewmodels.ItemViewModel$loadData$1.doLoad(Unknown Source:4)
05-07 23:08:05.443 W/System.err(31497): 	at jahirfiquitiva.libs.archhelpers.tasks.QAsync$execute$1.invoke(Unknown Source:23)
05-07 23:08:05.443 W/System.err(31497): 	at jahirfiquitiva.libs.archhelpers.tasks.QAsync$execute$1.invoke(Unknown Source:2)
05-07 23:08:05.443 W/System.err(31497): 	at org.jetbrains.anko.AsyncKt$doAsync$1.invoke(Unknown Source:4)
05-07 23:08:05.443 W/System.err(31497): 	at org.jetbrains.anko.AsyncKt$doAsync$1.invoke(Unknown Source:0)
05-07 23:08:05.443 W/System.err(31497): 	at org.jetbrains.anko.AsyncKt$sam$Callable$761a5578.call(Unknown Source:2)
05-07 23:08:05.443 W/System.err(31497): 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
05-07 23:08:05.443 W/System.err(31497): 	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:301)
05-07 23:08:05.443 W/System.err(31497): 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
05-07 23:08:05.443 W/System.err(31497): 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
05-07 23:08:05.443 W/System.err(31497): 	at java.lang.Thread.run(Thread.java:764)
```


```groovy
05-07 23:08:16.696 W/System.err(31592): java.util.zip.ZipException: ZIP_Read: error reading zip file
05-07 23:08:16.697 W/System.err(31592): 	at java.util.zip.ZipFile.read(Native Method)
05-07 23:08:16.697 W/System.err(31592): 	at java.util.zip.ZipFile.-wrap0(Unknown Source:0)
05-07 23:08:16.697 W/System.err(31592): 	at java.util.zip.ZipFile$ZipFileInputStream.read(ZipFile.java:758)
05-07 23:08:16.697 W/System.err(31592): 	at java.util.zip.ZipFile$ZipFileInflaterInputStream.fill(ZipFile.java:447)
05-07 23:08:16.697 W/System.err(31592): 	at java.util.zip.InflaterInputStream.read(InflaterInputStream.java:163)
05-07 23:08:16.697 W/System.err(31592): 	at java.io.FilterInputStream.read(FilterInputStream.java:107)
05-07 23:08:16.697 W/System.err(31592): 	at kotlin.io.ByteStreamsKt.copyTo(Unknown Source:27)
05-07 23:08:16.697 W/System.err(31592): 	at jahirfiquitiva.libs.kuper.helpers.extensions.FileKt.copyFromTo(Unknown Source:32)
05-07 23:08:16.697 W/System.err(31592): 	at jahirfiquitiva.libs.kuper.providers.viewmodels.KuperViewModel.getWidgetPreviewsPathFromZip(Unknown Source:175)
05-07 23:08:16.697 W/System.err(31592): 	at jahirfiquitiva.libs.kuper.providers.viewmodels.KuperViewModel.internalLoad(Unknown Source:224)
05-07 23:08:16.697 W/System.err(31592): 	at jahirfiquitiva.libs.kuper.providers.viewmodels.KuperViewModel.internalLoad(Unknown Source:2)
05-07 23:08:16.697 W/System.err(31592): 	at jahirfiquitiva.libs.archhelpers.viewmodels.ItemViewModel.safeInternalLoad(Unknown Source:15)
05-07 23:08:16.697 W/System.err(31592): 	at jahirfiquitiva.libs.archhelpers.viewmodels.ItemViewModel.access$safeInternalLoad(Unknown Source:0)
05-07 23:08:16.697 W/System.err(31592): 	at jahirfiquitiva.libs.archhelpers.viewmodels.ItemViewModel$loadData$1.doLoad(Unknown Source:4)
05-07 23:08:16.697 W/System.err(31592): 	at jahirfiquitiva.libs.archhelpers.tasks.QAsync$execute$1.invoke(Unknown Source:23)
05-07 23:08:16.697 W/System.err(31592): 	at jahirfiquitiva.libs.archhelpers.tasks.QAsync$execute$1.invoke(Unknown Source:2)
05-07 23:08:16.697 W/System.err(31592): 	at org.jetbrains.anko.AsyncKt$doAsync$1.invoke(Unknown Source:4)
05-07 23:08:16.697 W/System.err(31592): 	at org.jetbrains.anko.AsyncKt$doAsync$1.invoke(Unknown Source:0)
05-07 23:08:16.697 W/System.err(31592): 	at org.jetbrains.anko.AsyncKt$sam$Callable$761a5578.call(Unknown Source:2)
05-07 23:08:16.697 W/System.err(31592): 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
05-07 23:08:16.697 W/System.err(31592): 	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:301)
05-07 23:08:16.697 W/System.err(31592): 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
05-07 23:08:16.697 W/System.err(31592): 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
05-07 23:08:16.697 W/System.err(31592): 	at java.lang.Thread.run(Thread.java:764)
```

```groovy
05-07 23:08:16.785 W/System.err(31592): java.util.zip.ZipException: File too short to be a zip file: 0
05-07 23:08:16.786 W/System.err(31592): 	at java.util.zip.ZipFile.<init>(ZipFile.java:214)
05-07 23:08:16.786 W/System.err(31592): 	at java.util.zip.ZipFile.<init>(ZipFile.java:149)
05-07 23:08:16.786 W/System.err(31592): 	at java.util.zip.ZipFile.<init>(ZipFile.java:163)
05-07 23:08:16.786 W/System.err(31592): 	at jahirfiquitiva.libs.kuper.providers.viewmodels.KuperViewModel.getWidgetPreviewsPathFromZip(Unknown Source:140)
05-07 23:08:16.786 W/System.err(31592): 	at jahirfiquitiva.libs.kuper.providers.viewmodels.KuperViewModel.internalLoad(Unknown Source:224)
05-07 23:08:16.786 W/System.err(31592): 	at jahirfiquitiva.libs.kuper.providers.viewmodels.KuperViewModel.internalLoad(Unknown Source:2)
05-07 23:08:16.786 W/System.err(31592): 	at jahirfiquitiva.libs.archhelpers.viewmodels.ItemViewModel.safeInternalLoad(Unknown Source:15)
05-07 23:08:16.786 W/System.err(31592): 	at jahirfiquitiva.libs.archhelpers.viewmodels.ItemViewModel.access$safeInternalLoad(Unknown Source:0)
05-07 23:08:16.786 W/System.err(31592): 	at jahirfiquitiva.libs.archhelpers.viewmodels.ItemViewModel$loadData$1.doLoad(Unknown Source:4)
05-07 23:08:16.786 W/System.err(31592): 	at jahirfiquitiva.libs.archhelpers.tasks.QAsync$execute$1.invoke(Unknown Source:23)
05-07 23:08:16.786 W/System.err(31592): 	at jahirfiquitiva.libs.archhelpers.tasks.QAsync$execute$1.invoke(Unknown Source:2)
05-07 23:08:16.786 W/System.err(31592): 	at org.jetbrains.anko.AsyncKt$doAsync$1.invoke(Unknown Source:4)
05-07 23:08:16.786 W/System.err(31592): 	at org.jetbrains.anko.AsyncKt$doAsync$1.invoke(Unknown Source:0)
05-07 23:08:16.786 W/System.err(31592): 	at org.jetbrains.anko.AsyncKt$sam$Callable$761a5578.call(Unknown Source:2)
05-07 23:08:16.786 W/System.err(31592): 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
05-07 23:08:16.786 W/System.err(31592): 	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:301)
05-07 23:08:16.786 W/System.err(31592): 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
05-07 23:08:16.786 W/System.err(31592): 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
05-07 23:08:16.786 W/System.err(31592): 	at java.lang.Thread.run(Thread.java:764)
05-07 23:08:16.786 E/BitmapFactory(31592): Unable to decode stream: java.io.FileNotFoundException: /storage/emulated/0/Android/data/jahirfiquitiva.apps.kuper.sample/cache/KuperPreviews/FimboSimpleBar.png (No such file or directory)
05-07 23:08:16.787 W/System.err(31592): java.lang.IllegalStateException: BitmapFactory.decodeFile(preview.absolutePath) must not be null
05-07 23:08:16.787 W/System.err(31592): 	at jahirfiquitiva.libs.kuper.providers.viewmodels.KuperViewModel.getWidgetPreviewsPathFromZip(Unknown Source:389)
05-07 23:08:16.787 W/System.err(31592): 	at jahirfiquitiva.libs.kuper.providers.viewmodels.KuperViewModel.internalLoad(Unknown Source:224)
05-07 23:08:16.787 W/System.err(31592): 	at jahirfiquitiva.libs.kuper.providers.viewmodels.KuperViewModel.internalLoad(Unknown Source:2)
05-07 23:08:16.787 W/System.err(31592): 	at jahirfiquitiva.libs.archhelpers.viewmodels.ItemViewModel.safeInternalLoad(Unknown Source:15)
05-07 23:08:16.787 W/System.err(31592): 	at jahirfiquitiva.libs.archhelpers.viewmodels.ItemViewModel.access$safeInternalLoad(Unknown Source:0)
05-07 23:08:16.787 W/System.err(31592): 	at jahirfiquitiva.libs.archhelpers.viewmodels.ItemViewModel$loadData$1.doLoad(Unknown Source:4)
05-07 23:08:16.787 W/System.err(31592): 	at jahirfiquitiva.libs.archhelpers.tasks.QAsync$execute$1.invoke(Unknown Source:23)
05-07 23:08:16.787 W/System.err(31592): 	at jahirfiquitiva.libs.archhelpers.tasks.QAsync$execute$1.invoke(Unknown Source:2)
05-07 23:08:16.787 W/System.err(31592): 	at org.jetbrains.anko.AsyncKt$doAsync$1.invoke(Unknown Source:4)
05-07 23:08:16.787 W/System.err(31592): 	at org.jetbrains.anko.AsyncKt$doAsync$1.invoke(Unknown Source:0)
05-07 23:08:16.787 W/System.err(31592): 	at org.jetbrains.anko.AsyncKt$sam$Callable$761a5578.call(Unknown Source:2)
05-07 23:08:16.787 W/System.err(31592): 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
05-07 23:08:16.788 W/System.err(31592): 	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:301)
05-07 23:08:16.788 W/System.err(31592): 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
05-07 23:08:16.788 W/System.err(31592): 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
05-07 23:08:16.788 W/System.err(31592): 	at java.lang.Thread.run(Thread.java:764)
```

**Comment by SanchithHegde** [comment]
Okay, no problem. I can wait.

**Comment by jahirfiquitiva** [comment]
@SanchithHegde Can you please send me the widgets you are trying to use, just so I can properly test the fix? Send them to [hi@jahirfiquitiva.com](mailto:hi@jahirfiquitiva.com]

**Comment by jahirfiquitiva** [comment]
Fixed in v `1.3.2`

**Comment by SanchithHegde** [comment]
Thank you!
